processing/src/main/java/org/apache/druid/query/aggregation/firstlast/FirstLastVectorAggregator.java

```diff
@@ -0,0 +1,271 @@
+      final boolean[] valueNullityVector = valueSelector.getNullVector();
+      Integer selectedIndex = null;
+
+      for (int index = startRow; index < endRow; ++index) {
+        if (timeNullityVector != null && timeNullityVector[index]) {
+          // Don't aggregate values where time isn't present
+          continue;
+        }
+        // Find the latest time inside the vector objects
+        if (selectedIndex == null) {
+          selectedIndex = index;
+        } else {
+          if (selectionPredicate.apply(timeVector[index], timeVector[selectedIndex])) {
+            selectedIndex = index;
+          }
+        }
+      }
+      // Compare the selectedIndex's value to the value on the buffer. This way, we write to the buffer only once
+      // Weeds out empty vectors, where endRow == startRow
+      if (selectedIndex != null) {
+        if (selectionPredicate.apply(timeVector[selectedIndex], buf.getLong(position))) {
+          // Write the value here
+          if (valueNullityVector == null || !valueNullityVector[selectedIndex]) {
+            putValue(buf, position, timeVector[selectedIndex], valueSelector, selectedIndex);
+          } else if (useDefault) {
+            putDefaultValue(buf, position, timeVector[selectedIndex]);
+          } else {
+            putNull(buf, position, timeVector[selectedIndex]);
+          }
+        }
+      }
+    }
+  }
+
+  @Override
+  public void aggregate(
+      ByteBuffer buf,
+      int numRows,
+      int[] positions,
+      @Nullable int[] rows,
+      int positionOffset
+  )
+  {
+    // Not a normal case, and this doesn't affect the folding. timeSelectors should be present (albeit irrelevent) when folding.
+    // timeSelector == null means that the aggregating column's capabilities aren't known, and it only happens for a special case
+    // while building string aggregator
+    if (timeSelector == null) {
+      return;
+    }
+
+    // If objectSelector isn't null, then the objects might be folded up. If that's the case, whatever's represented by
+    // the timeSelector doesn't hold any relevance. We should check for folded objects BEFORE even thinking about looking
+    // at the timeSelector
+    if (objectSelector != null) {
+      final Object[] maybeFoldedObjects = objectSelector.getObjectVector();
+      final boolean[] timeNullityVector = timeSelector.getNullVector();
+      final long[] timeVector = timeSelector.getLongVector();
+      for (int i = 0; i < numRows; ++i) {
+        int position = positions[i] + positionOffset;
+        int row = rows == null ? i : rows[i];
+        long lastTime = buf.getLong(position);
+        // All the information about the object would be in the single selector. This method will check the folding of the object selector,
+        // casting, and everything........
+        PairType pair = readPairFromVectorSelectors(timeNullityVector, timeVector, maybeFoldedObjects, row);
+        if (pair != null) {
```

Comment:
note: this conditional may make the computation to get the `lastTime` field from the `buf` (random read) unnecessary...not sure how frequently that may happen...

Reply:
Good point, it can be optimised.